File: bst/cpp/lb-09-convert-binary-tree-to-bst.cpp
void find_preorder(Node *root, vector<int> &arr) {
    if (!root) return;
    find_preorder(root->left, arr);
    arr.push_back(root->data);
    find_preorder(root->right, arr);
}
void place(Node *root, int &i, vector<int> &pre) {
    if (!root) return;
    place(root->left, i, pre);
    root->data = pre[i++];
    place(root->right, i, pre);
}
// The given root is the root of the Binary Tree
// Return the root of the generated BST
Node *binaryTreeToBST(Node *root) {
    vector<int> pre;
    find_preorder(root, pre);
    sort(pre.begin(), pre.end());
    int i = 0;
    place(root, i, pre);
    return root;
}

File: bst/lb-09-convert-binary-tree-to-bst.py
# You need to modify binary tree to bst without chaging the structure.
# Find inorder of binary tree
# Sort inorder
# traverse binary tree in inorder and place the values from sorted array.

def convert(root):
    inor = []

    def inorder(root):
        if not root:
            return
        inorder(root.left)
        inor.append(root.data)
        inorder(root.right)

    inorder(root)
    # sort
    sorted_inor = sorted(inor)
    # palce the value
    i = 0

    def place_inorder(root):
        nonlocal i
        if not root:
            return
        place_inorder(root.left)
        root.data = sorted_inor[i]
        i += 1
        place_inorder(root.right)

    place_inorder(root)
